### ISSUE TITLE
Second edition: page 127, page 141

### ISSUE THREAD

**Issue by aixtools** [geerlingguy/ansible-for-devops#450]
- Reading `ansible-doc` and other things I am (now) much more aware of the idea of "control" versus "remote".
- other than the string in the index, page 127 is the first time the _attribute_ `connection:` comes up.

- What is the point (and impact) of
```
connection: local
```

Page: 141

I like the idea of `include_tasks`, but I would also like the _safety-net_ like on page 127

would something like:
```
- name: "Check for {{ username }} customization."
  stat:
    path: "tasks/{{ username }}.yml"
  register: customization
  connection: local

- name: "{Perform {{ username }} customization"
  include_tasks: "{{ username }}.yml"
  when: customization.stat.exists
```

And is using **connection: local**, useful, wrong, or just pretty?

- Intent: A reader trying to apply what has just been introduced earlier - but unsure what the default relative path is during execution of  `.../tasks/main.yml`.

**Comment by aixtools** [comment]
Maybe off-topic, but as it relates to my first attempt at compressing a number of long playbooks - written as "roles", but basically, in many locations duplicates of each other.

The "users" `tasks/main.yml` is now:
```
---
- name: "Create {{ username }} account"
  command:
    cmd: /usr/bin/mkuser
      home="/home/{{ username }}"
      shell="/usr/bin/ksh"
      "{{ username }}"
    creates: "/home/{{ username }}"

- name: "Create {{ username }} .ssh directory"
  file:
    path: /home/{{ username }}/.ssh
    owner: "{{ username }}"
    group: staff
    mode: 0700
    state: directory

- name: "Set/reset {{ username }} limits"
  command:
    cmd: /usr/bin/chuser
      "{{ limits.attr }}={{ limits.value | default('-1') }}"
      "{{ username }}"
  loop:
    "{{ user_limits }}"
  loop_control:
    loop_var: limits

- name: "Check for {{ username }} customization."
  stat:
    path: "tasks/{{ username }}.yml"
  register: customization
  connection: local

- name: "{Perform {{ username }} customization"
  include_tasks: "{{ username }}.yml"
  when: customization.stat.exists
```

And works when given a _static_ argument. (emphasis on _static_) - as I found the answer elsewhere - and you talk about the difference of `include_*` versus `import_*` - and if that was meant to clarify that `roles` are static calls while `tasks:` are dynamic - I missed that - that is my _feedback_.

The problem I was having was I wrote `main.yml` as:
```
---
- hosts: wip
  gather_facts: no
  # roles are 'static' so they do not work with with_items.

  roles:
    - role: users
      username: "tim"
    - role: users
      username: "{{ item }}"
      with_items:
        - "test"
        - "user2"
```
* imho: ansible documentation does a poor job of saying loop iteration is permitted.
* However, I found a statement at the bottom of this question/answer: https://groups.google.com/g/ansible-project/c/VtYw0Pzgr0o?pli=1
```
Brian Coca
unread,
Jun 2, 2017, 5:20:35 PM
to Ansible Project
Well, roles: have always been statick, with_ and other directives ONLY
execute on tasks, so when adding them to static imports (like roles:)
they should just be inherited by the imported tasks.

----------
Brian Coca 
```
* so, feedback again: this is, imho, a good place to emphasize that _iteration_ only works on a certain type of (searching for what I hope is the right word - I'll use **directive**) ansible _directives_ only work in **tasks:**.

* Working example of above - using both _roles:_ and _tasks:_
```
# roles: are 'static' so they do not work with with_items.
# tasks: work with_* iteration. Coupled with 'include_role', with_* is possible
# vars: will be moved to either of:
#  .../vars/main.yml
#  .../default/main.yml
# kept here for simplicity atm
---
- hosts: wip
  gather_facts: no
  vars:
    userlist:
      - "test"
      - "user2"

  roles:
    - role: users
      username: "tim"
  tasks:
    - name: "Manage userlist"
      include_role:
        name: users
      vars:
        username: "{{ item }}"
      with_items:
        "{{ userlist }}"
```

**Comment by aixtools** [comment]
- a bit more head-spinning (the limitations of what can be included in a role "tasks" playbook confound me.
- You give an example - but I guess it is just before you start with roles - and I seem to be expected to forget many of the things just mentioned:
- e.g., in the sample below - I am trying to dynamically expand the role based on a task file being present in the role tasks directory.
- the debug statement tells me the 'exists' is always false, even when the file exists - so I guess - my oath: is not correct.
- please feedback (aka help):
```
- name: "Check for {{ username }} customization."
  stat:
    path: "tasks/{{ username }}.yml"
  register: customization
#  connection: local
  loop:
    "{{ user_names }}"
  loop_control:
    loop_var: username

- name: "Debug"
  debug:
    verbosity: 0
    var: customization

- name: "{Perform {{ cust.username }} customization"
  include_tasks: "{{ cust.username }}.yml"
  when: cust.stat.exists }}"
  loop: "{{ customization.results }}"
  loop_control:
    loop_var: cust
```

**Comment by geerlingguy** [comment]
Thanks for all these notes—the difficulty is it's hard to parse them all out as one large issue to try to find any actionable changes in the book. If you can maybe split things into separate issues, that would be helpful in getting any changes made.
